apps/script/HoC2023AiGenerateWeights.py

```diff
@@ -107,33 +111,83 @@
+embedding_paths = ['apps/static/dance/ai/model/input_embeddings.pkl',
+'apps/static/dance/ai/model/palette_embeddings.pkl',
+'apps/static/dance/ai/model/background_embeddings.pkl',
+'apps/static/dance/ai/model/foreground_embeddings.pkl']
+input_path = embedding_paths[0]
+palette_path = embedding_paths[1]
+background_path = embedding_paths[2]
+foreground_path = embedding_paths[3]
+
+def load_embedding_cache(path):
+    # load the cache if it exists, and save a copy to disk
+    try:
+        embedding_cache = pd.read_pickle(path)
+    except FileNotFoundError:
+        embedding_cache = {}
+    with open(path, "wb") as embedding_cache_file:
+        pickle.dump(embedding_cache, embedding_cache_file)
+    return embedding_cache
+
+caches = [load_embedding_cache(path) for path in embedding_paths]
+input_cache = caches[0]
+palette_cache = caches[1]
+background_cache = caches[2]
+foreground_cache = caches[3]
+
+# Define a function to retrieve embeddings from the cache if present, and otherwise request via the API
+def retrieve_embedding(string: str,
+    cache_path: str,
+    embedding_cache,
+    model: str = EMBEDDING_MODEL,
+) -> list:
+    """Return embedding of given string, using a cache to avoid recomputing."""
+    if (string, model) not in embedding_cache.keys():
+        embedding_cache[(string, model)] = get_embedding(string, model)
+        with open(cache_path, "wb") as embedding_cache_file:
+            pickle.dump(embedding_cache, embedding_cache_file)
+    return embedding_cache[(string, model)]
+
+# Retrieve embeddings for input emojis, palettes, backgrounds, and foregrounds in that order
+option_lists = [emoji_names_list, palette_model_descriptive_names, background_effect_model_descriptive_names, foreground_effect_model_descriptive_names]
+embeddings = []
+# Final output should be list of lists structured as [[[emoji1], [emoji2], ...], [[palette1]...]...]
+# Where embeddings[0] = all emoji embeddings, [1] = palettes, [2] = background, [3] = foreground
+for i in range (0, 4):
+    embeddings.append([retrieve_embedding(string=item, 
+                                          cache_path=embedding_paths[i],
+                                          embedding_cache=caches[i])
+                                          for item in option_lists[i]])
+
+emoji_embeddings = embeddings[0]
+palette_embeddings = embeddings[1]
+background_embeddings = embeddings[2]
+foreground_embeddings = embeddings[3]
+
+def calculate_similarity_score(input_embeddings, output_embeddings):
+    # Native cosine distance calculation outputs a value between 0 -> 1 where smaller values = greater similarity
+    # We can redefine this into cosine similarity with a simple (x-1)*-1 due to their mathematical relationship
+    # Cosine similarity is preferable as we can easily sum them together to take a max value later
+    similarities = [distances_from_embeddings(input_vector, output_embeddings, distance_metric='cosine')
```

Comment:
I read that others were encountering having similarity scores that seemed clustered around 0.7 [thread](https://community.openai.com/t/some-questions-about-text-embedding-ada-002-s-embedding/35299). Someone suggested using euclidean distance. Maybe we could do a comparison? 
We could try `distance_metric='L2'`.
https://github.com/openai/openai-python/blob/284c1799070c723c6a553337134148a7ab088dd8/openai/embeddings_utils.py#L145

Reply:
We discussed offline and Tyrone pointed out that euclidean distance can also take into account length of the string which is undesirable.